be/src/runtime/mem_tracker.cpp

```diff
@@ -242,26 +243,38 @@ int64_t MemTracker::GetPoolMemReserved() {
     return mem_reserved;
 }
 
-std::shared_ptr<MemTracker> PoolMemTrackerRegistry::GetRequestPoolMemTracker(
-        const string& pool_name, bool create_if_not_present) {
-    DCHECK(!pool_name.empty());
-    lock_guard<SpinLock> l(pool_to_mem_trackers_lock_);
-    PoolTrackersMap::iterator it = pool_to_mem_trackers_.find(pool_name);
-    if (it != pool_to_mem_trackers_.end()) {
-        MemTracker* tracker = it->second.get();
-        DCHECK(pool_name == tracker->pool_name_);
-        return it->second;
+std::shared_ptr<MemTracker> QueryMemTrackerRegistry::RegisterQueryMemTracker(
+        const std::string& query_id, int64_t mem_limit) {
+    DCHECK(!query_id.empty());
+    if (mem_limit != -1) {
+        if (mem_limit > MemInfo::physical_mem()) {
+            LOG(WARNING) << "Memory limit " << PrettyPrinter::print(mem_limit, TUnit::BYTES)
+                         << " exceeds physical memory of "
+                         << PrettyPrinter::print(MemInfo::physical_mem(), TUnit::BYTES);
+        }
+        VLOG(2) << "Using query memory limit: " << PrettyPrinter::print(mem_limit, TUnit::BYTES);
     }
-    if (!create_if_not_present) return nullptr;
-    // First time this pool_name registered, make a new object.
-    std::shared_ptr<MemTracker> tracker = MemTracker::CreateTracker(
-            -1, strings::Substitute(REQUEST_POOL_MEM_TRACKER_LABEL_FORMAT, pool_name),
-            ExecEnv::GetInstance()->process_mem_tracker());
-    tracker->pool_name_ = pool_name;
-    pool_to_mem_trackers_.emplace(pool_name, std::shared_ptr<MemTracker>(tracker));
+
+    // First time this query_id registered, make a new object, otherwise do nothing.
+    _query_mem_trackers.try_emplace_l(
+            query_id, [](std::shared_ptr<MemTracker>) {},
+            MemTracker::CreateTracker(mem_limit,
+                                      strings::Substitute(QUERY_MEM_TRACKER_LABEL_FORMAT, query_id),
+                                      ExecEnv::GetInstance()->hook_process_mem_tracker(), false,
+                                      false, MemTrackerLevel::OVERVIEW, query_id));
+
+    std::shared_ptr<MemTracker> tracker = nullptr;
+    _query_mem_trackers.if_contains(query_id,
+                                    [&tracker](std::shared_ptr<MemTracker> v) { tracker = v; });
     return tracker;
 }
 
+void QueryMemTrackerRegistry::DeregisterQueryMemTracker(const std::string& query_id) {
+    DCHECK(!query_id.empty());
+    _query_mem_trackers.erase_if(query_id, [](std::shared_ptr<MemTracker>) { return true; });
+    LOG(WARNING) << "DeregisterQueryMemTracker " << query_id << " len " << _query_mem_trackers.size();
```

Comment:
Why there is a warning log?

Reply:
sorry, This is temp log, forgot to delete it =_=